src/components/smartSearch/EditTargetDialog.tsx

```diff
@@ -0,0 +1,128 @@
+import { FormattedMessage as Msg } from 'react-intl';
+import { useState } from 'react';
+import { Box, Button, ButtonBase, Card, CardContent, Dialog, DialogContent, Typography } from '@material-ui/core';
+
+import All from './filters/All';
+import Filter from './Filter';
+import { isFilterWithId } from './utils';
+import MostActive from './filters/MostActive';
+import patchTaskTarget from 'fetching/tasks/patchTaskTarget';
+import { useRouter } from 'next/router';
+import { ZetkinSmartSearchFilter } from 'types/zetkin';
+import { FILTER_TYPE, ZetkinSmartSearchFilterWithId } from 'types/smartSearch';
+import { useMutation, useQueryClient } from 'react-query';
+
+interface EditTargetDialogProps {
+    filterSpec: ZetkinSmartSearchFilter[];
+    onDialogClose: () => void;
+    open: boolean;
+}
+
+const EditTargetDialog = ({ onDialogClose, open, filterSpec }: EditTargetDialogProps) : JSX.Element => {
+    const queryClient = useQueryClient();
+    const { orgId, taskId } = useRouter().query;
+    const filtersWithIds = filterSpec.map((filter, index) => ({ ...filter, id: index }));
+    const [filterArray, setFilterArray] = useState<ZetkinSmartSearchFilterWithId[]>(filtersWithIds);
+    const [selectedType, setSelectedType] = useState<FILTER_TYPE | null>(null);
+    const [selectedFilter, setSelectedFilter] = useState<ZetkinSmartSearchFilterWithId| null>(null);
+
+    const taskMutation = useMutation(patchTaskTarget(orgId as string, taskId as string),{
+        onSettled: () => queryClient.invalidateQueries(['filter_spec', orgId, taskId]),
+    } );
+
+    const handleDialogClose = () => {
+        setSelectedType(null);
+        onDialogClose();
+    };
+
+    const handleCancelFilter = () => setSelectedType(null);
+
+    const handleSubmitFilter = (filter:ZetkinSmartSearchFilterWithId | ZetkinSmartSearchFilter) => {
+        if (isFilterWithId(filter)) {
+            setFilterArray(filterArray.map(f => {
+                if (f.id === filter.id) {
+                    return filter;
+                }
+                else {
+                    return f;
+                }
+            }));
+        }
+        else {
+            const newId = 1 + (filterArray.length ? Math.max(...filterArray.map(f => f.id)) : 0);
+            setFilterArray(filterArray.concat({ ...filter, id: newId }));
+        }
+        setSelectedFilter(null);
+        setSelectedType(null);
+    };
+
+    const handleSelectedTypeChange = (type: FILTER_TYPE) => setSelectedType(type);
```

Comment:
If we wanted to get rid of the derived `selectedType` state, you could `setSelectedFilter({ type }) here, i.e. create a new filter containing nothing but a type and pass it into `setSelectedFilter`.

This could in fact be used to generate IDs as well, which might simplify the current approach to IDs?

Reply:
addressed in #310